controllers/note.py

```diff
@@ -0,0 +1,107 @@
+import math
+
+from magicbot import StateMachine, state, timed_state, will_reset_to
+from wpimath.geometry import Translation2d
+
+from components.chassis import ChassisComponent
+from components.injector import InjectorComponent
+from components.intake import IntakeComponent
+from components.shooter import ShooterComponent
+from utilities.game import get_goal_speaker_position
+
+
+class NoteManager(StateMachine):
+
+    shooter_component: ShooterComponent
+
+    chassis: ChassisComponent
+    injector_component: InjectorComponent
+    intake: IntakeComponent
+
+    shot_desired = will_reset_to(False)
+    intake_desired = will_reset_to(False)
+    intake_cancel_desired = will_reset_to(False)
+
+    def try_intake(self):
+        self.intake_desired = True
+
+    def cancel_intake(self):
+        self.intake_cancel_desired = True
+
+    def try_shoot(self):
+        self.shot_desired = True
+
+    def has_note(self):
+        return self.injector_component.has_note()
+
+    @property
+    def translation_to_goal(self) -> Translation2d:
+        return (
+            get_goal_speaker_position().toTranslation2d()
+            - self.chassis.get_pose().translation()
+        )
+
+    @state(must_finish=True)
+    def idling(self):
+        self.intake.retract()
+
+        # Update range
+        self.shooter_component.set_range(self.translation_to_goal.norm())
+
+        if self.intake_desired:
+            self.next_state(self.dropping_intake)
+
+    @state(must_finish=True)
+    def dropping_intake(self):
+        self.intake.deploy()
+
+        # Update range
+        self.shooter_component.set_range(self.translation_to_goal.norm())
+
+        if self.intake.is_fully_deployed():
+            self.next_state(self.intaking)
+
+    @state(must_finish=True)
+    def intaking(self):
+        self.intake.intake()
+        self.injector_component.intake()
+
+        # Update range
+        self.shooter_component.set_range(self.translation_to_goal.norm())
+
+        if self.injector_component.has_note():
+            self.next_state(self.holding_note)
+
+        if self.intake_cancel_desired:
+            self.next_state(self.idling)
+
+    @state(first=True, must_finish=True)
+    def holding_note(self):
```

Comment:
This might be correct for auto, but will this always be correct at the start of teleop?